# Autoencoder

In [14]:
import os

from utils.loaders import load_mnist
from models.AE import Autoencoder

## Set parameters

In [15]:
# run params
SECTION = 'vae'
RUN_ID = '0001'
DATA_NAME = 'digits'
RUN_FOLDER = 'run/{}/'.format(SECTION)
RUN_FOLDER += '_'.join([RUN_ID, DATA_NAME])

print(RUN_FOLDER)

if not os.path.exists(RUN_FOLDER):
    os.mkdir(RUN_FOLDER)
    os.mkdir(os.path.join(RUN_FOLDER, 'viz'))
    os.mkdir(os.path.join(RUN_FOLDER, 'images'))
    os.mkdir(os.path.join(RUN_FOLDER, 'weights'))

MODE =  'build' #'load' #

run/vae/0001_digits


## Load the data

In [16]:
(x_train, y_train), (x_test, y_test) = load_mnist()

## Define the structure of the neural network

In [17]:
AE = Autoencoder(
    input_dim = (28,28,1)
    , encoder_conv_filters = [32,64,64,64]  # conv層の出力フィルタ数 conv層は合計4層
    , encoder_conv_kernel_size = [3,3,3,3]  # conv層のカーネル（フィルタ）サイズ
    , encoder_conv_strides = [1,2,2,1]  # convのstride数 1:出力サイズ同じ, 2:出力サイズ半分
    , decoder_conv_t_filters = [64,64,32,1]
    , decoder_conv_t_kernel_size = [3,3,3,3]
    , decoder_conv_t_strides = [1,2,2,1]
    , z_dim = 2
)

if MODE == 'build':
    AE.save(RUN_FOLDER)
else:
    AE.load_weights(os.path.join(RUN_FOLDER, 'weights/weights.h5'))

In [18]:
AE.encoder.summary()

Model: "model_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
encoder_input (InputLayer)   [(None, 28, 28, 1)]       0         
_________________________________________________________________
encoder_conv_0 (Conv2D)      (None, 28, 28, 32)        320       
_________________________________________________________________
leaky_re_lu_7 (LeakyReLU)    (None, 28, 28, 32)        0         
_________________________________________________________________
encoder_conv_1 (Conv2D)      (None, 14, 14, 64)        18496     
_________________________________________________________________
leaky_re_lu_8 (LeakyReLU)    (None, 14, 14, 64)        0         
_________________________________________________________________
encoder_conv_2 (Conv2D)      (None, 7, 7, 64)          36928     
_________________________________________________________________
leaky_re_lu_9 (LeakyReLU)    (None, 7, 7, 64)          0   

encoder_conv_0 (Conv2D)    Param:320<br>
(3 * 3 * 1 + 1) * 32 = 320<br>
(filter h  ×  filter w  ×  input channel  +  bias)  ×  output channel<br><br>
encoder_conv_1 (Conv2D)    Param:18496<br> 
(3 * 3 * 32 + 1) * 64 = 18496



↑最終的には2次元ベクトルにエンコードしている

In [19]:
AE.decoder.summary()

Model: "model_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
decoder_input (InputLayer)   [(None, 2)]               0         
_________________________________________________________________
dense_1 (Dense)              (None, 3136)              9408      
_________________________________________________________________
reshape_1 (Reshape)          (None, 7, 7, 64)          0         
_________________________________________________________________
decoder_conv_t_0 (Conv2DTran (None, 7, 7, 64)          36928     
_________________________________________________________________
leaky_re_lu_11 (LeakyReLU)   (None, 7, 7, 64)          0         
_________________________________________________________________
decoder_conv_t_1 (Conv2DTran (None, 14, 14, 64)        36928     
_________________________________________________________________
leaky_re_lu_12 (LeakyReLU)   (None, 14, 14, 64)        0   

In [20]:
 # エンコーダとデコーダを連結する
AE.model.summary()

Model: "model_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
encoder_input (InputLayer)   [(None, 28, 28, 1)]       0         
_________________________________________________________________
encoder_conv_0 (Conv2D)      (None, 28, 28, 32)        320       
_________________________________________________________________
leaky_re_lu_7 (LeakyReLU)    (None, 28, 28, 32)        0         
_________________________________________________________________
encoder_conv_1 (Conv2D)      (None, 14, 14, 64)        18496     
_________________________________________________________________
leaky_re_lu_8 (LeakyReLU)    (None, 14, 14, 64)        0         
_________________________________________________________________
encoder_conv_2 (Conv2D)      (None, 7, 7, 64)          36928     
_________________________________________________________________
leaky_re_lu_9 (LeakyReLU)    (None, 7, 7, 64)          0   

## Train the autoencoder

In [21]:
LEARNING_RATE = 0.0005
BATCH_SIZE = 128
INITIAL_EPOCH = 0

In [22]:
AE.compile(LEARNING_RATE)

In [23]:
AE.train(
    x_train[:1000]
    , batch_size = BATCH_SIZE
    , epochs = 200
    , run_folder = RUN_FOLDER
    , initial_epoch = INITIAL_EPOCH
)

=============================] - 1s 110ms/step - loss: 0.0453 - lr: 5.0000e-04
Epoch 114/200
8/8 [==============================] - ETA: 0s - loss: 0.0452
Epoch 00114: saving model to run/vae/0001_digits/weights/weights.h5
8/8 [==============================] - 1s 104ms/step - loss: 0.0452 - lr: 5.0000e-04
Epoch 115/200
8/8 [==============================] - ETA: 0s - loss: 0.0451
Epoch 00115: saving model to run/vae/0001_digits/weights/weights.h5
8/8 [==============================] - 1s 103ms/step - loss: 0.0451 - lr: 5.0000e-04
Epoch 116/200
8/8 [==============================] - ETA: 0s - loss: 0.0451
Epoch 00116: saving model to run/vae/0001_digits/weights/weights.h5
8/8 [==============================] - 1s 103ms/step - loss: 0.0451 - lr: 5.0000e-04
Epoch 117/200
8/8 [==============================] - ETA: 0s - loss: 0.0451
Epoch 00117: saving model to run/vae/0001_digits/weights/weights.h5
8/8 [==============================] - 1s 102ms/step - loss: 0.0451 - lr: 5.0000e-04
Epoch